README.md
# LXHelper

[![Build Status](https://drone.osshelp.ru/api/badges/OSSHelp/lxhelper/status.svg)](https://drone.osshelp.ru/OSSHelp/lxhelper)

## About

This script is used with the lxd-functions library for LXD container deployment.

## Requirements

* Ubuntu Xenial/Bionic/Focal
* LXD client and accessible LXD server (2.15+, 3.x, 4.x)

There is also limited support for the following distributions:

* CentOS 7.x or newer
* Debian 8.x or newer

Mostly it should work but without any guarantees ;)

## How to use it

### Installtion

LXHelper is just a wrapper around lxd-functions.sh library. For deployment you need:

1. Place lxd-function.sh library in the `/usr/local/include/osshelp/` path
1. Place lxhelper script in the `/usr/local/sbin/` path
1. Place the container YAML config in the `/usr/local/etc/lxc/` path (see EXAMPLES.md)

There’re install/update scripts in the repository. Command for installation (lxd-function.sh, lxhelper, yq and minio-client):

```shell
curl -s https://oss.help/scripts/lxc/lxhelper/install.sh | bash

### Options

LXHelper usage is described in the table below:

Action | Description| Usage
---|---|---
deploy-container |deploy one container from specified YAML config |`lxhelper deploy-container [-y=(URL|Path)] [--force] container_name`
deploy-containers |deploy all containers from specified YAML config | `lxhelper deploy-containers [-y=(URL|Path)] [--force]`
download-image|download container image with specified source URL in the YAML config |`lxhelper download-image [-y=(URL|Path)] container_name`
create-image|create image from the container with the same name or with specified name |`lxhelper create-image container_name [image_name]`
export-image|export specified image to archive|`lxhelper export-image image_name`
update-image|create and export container image from the container |`lxhelper update-image image_name`
update-configs|update all YAML configs in specified path (default path /usr/local/etc/lxc/) |`lxhelper update-configs [/path/to/dir]`

```

### Useful

LXHelper is tightly coupled with [deploy-functions](https://github.com/OSSHelp/deploy-functions) library, but it's still possible to use LXHelper without it.

## Author

OSSHelp Team, see <https://oss.help>


CHANGELOG.md
# Changelog

## v1.38

* Added postdeploy tests support

## v1.37

* Fixed containers limiting bug (when profiles aren't exists)

## v1.35

* Added limiting support for containers to profiles

## v1.33

* Added LXD version check (minimum required version is 2.15)

## v1.32

* Added name, network and devices parameters supports to profiles

## v1.31

* Added used profile name pushing into the container for default-setup.local

## v1.23

* Added files download feature (for download section in the container yaml configuration file), works with S3 storage and HTTP protocol with authorization

## v1.22

* Added container image download feature without image import

## v1.17

* Added protection for containers which weren't deployed by lxhelper (lxhelper checks key config.user.deployed_from in the containers configs)

## v1.15

* Fixed warning when cpu.allowance was used in the container config
* Updated warning messages
* Added yaml syntax test
* Added check for null values in keys in the yaml config

## v1.14

* Added warning for absented CPU/RAM limits in the container config

## v1.10

* Fixed bug in push_profile function

## v1.09

* Fixed bug when used limits with a dot in the name (e.g. cpu.allowance)
* Added drives limiting feature (mounting folders)

## v1.08

* Fixed bug with broken permissions for mounted directories (remade mount_directories_to_container function)
* Optimized information messages output for mount_directories_to_container function call


lxhelper/install.sh
#!/bin/bash
# shellcheck disable=SC2015

umask 0077
export LANG=C
export LC_ALL=C

declare -r lxdf_ver=1.42
declare -r lxdh_ver=1.07
silent=no
tmpdir="${TEMP:=/tmp}"

index_url="${REMOTE_URI:-https://oss.help/scripts/lxc/lxhelper/.list}"
list_name=$(basename "${index_url}")
script_name="lxhelper"
script_path="/usr/local/sbin"
func_name="lxd-functions.sh"
func_path="/usr/local/include/osshelp"
yq_bin='yq'
yq_bin_path='/usr/local/bin'
minio_client_bin='minio-client'
minio_client_path='/usr/local/sbin'
shacmd=$(command -v sha256sum || command -v gsha256sum 2>/dev/null)
err=0

function show_notice() { test "${silent}" != "yes" && echo -e "[NOTICE] ${*}"; return 0; }
function show_error() { echo -e "[ERROR] ${*}" >&2;  err=1; return 1; }

function fetch_files() {
  cd "${1}" && \
    {
      wget -q -P "${1}" "${index_url}" && \
      wget -q -i "${1}/${list_name}" -P "${1}"
    } && \
      ${shacmd} -c --status SHA256SUMS 2> /dev/null || {
        show_error "Something went wrong, checksums of downloaded files mismatch."
        ${shacmd} -c "${1}/SHA256SUMS"
        return 1
      }
}

function install_files() {
  test -d "${script_path}" || mkdir "${script_path}"
  cd "${script_path}" && \
    mv "${tmp_dir}/${script_name}" "${script_path}/${script_name}" && \
      chmod 700 "${script_path}/${script_name}"

  test -d "${func_path}" || mkdir "${func_path}"
  cd "${func_path}" && \
    mv "${tmp_dir}/${func_name}" "${func_path}/${func_name}" && \
      chmod 600 "${func_path}/${func_name}"

  test -d "${minio_client_path}" || mkdir "${minio_client_path}"
  cd "${minio_client_path}" && \
    mv "${tmp_dir}/${minio_client_bin}" "${minio_client_path}/${minio_client_bin}" && \
      chmod 755 "${minio_client_path}/${minio_client_bin}"
}

test "$(id -u)" != "0" && { show_error "Sorry, but you must run this script as root."; exit 1; }

tmp_dir="${tmpdir}/lxd-functions.$$"
mkdir -p "${tmp_dir}" && \
  fetch_files "${tmp_dir}" && \
    install_files "${tmp_dir}" && \
      {
        test -x "${yq_bin_path}/${yq_bin}" || curl -s https://oss.help/scripts/tools/yq/install.sh | bash
        show_notice "Minio-client was installed to ${minio_client_path}."
        show_notice "Library lxd-functions.sh (v${lxdf_ver}) was installed to $func_path."
        show_notice "Script lxhelper (v${lxdh_ver}) was installed to $script_path."
      }
test -d "${tmp_dir}" && rm -rf "${tmp_dir}"
test "${err}" -eq 1 && { show_error "Installation failed."; }
exit "${err}"


lxhelper/update.sh
#!/bin/bash
# shellcheck disable=SC2015

umask 0077
export LANG=C
export LC_ALL=C

declare -r lxdf_ver=1.42
declare -r lxdh_ver=1.07
silent=no
tmpdir="${TEMP:=/tmp}"

index_url="${REMOTE_URI:-https://oss.help/scripts/lxc/lxhelper/.list}"
list_name=$(basename "${index_url}")
script_name="lxhelper"
script_path="/usr/local/sbin"
func_name="lxd-functions.sh"
func_path="/usr/local/include/osshelp"
yq_bin='yq'
yq_bin_path='/usr/local/bin'
minio_client_bin='minio-client'
minio_client_path='/usr/local/sbin'
shacmd=$(command -v sha256sum || command -v gsha256sum 2>/dev/null)
err=0

function show_notice() { test "${silent}" != "yes" && echo -e "[NOTICE] ${*}"; return 0; }
function show_error() { echo -e "[ERROR] ${*}" >&2; err=1; return 1; }

function fetch_files() {
  cd "${1}" && \
    {
      wget -q -P "${1}" "${index_url}" && \
      wget -q -i "${1}/${list_name}" -P "${1}"
    } && \
      ${shacmd} -c --status SHA256SUMS 2> /dev/null || {
        show_error "Something went wrong, checksums of downloaded files mismatch."
        ${shacmd} -c "${1}/SHA256SUMS"
        return 1
      }
}

function install_files() {
  test -d "${script_path}" || mkdir "${script_path}"
  cd "${script_path}" && \
    mv "${tmp_dir}/${script_name}" "${script_path}/${script_name}" && \
      chmod 700 "${script_path}/${script_name}"

  test -d "${func_path}" || mkdir "${func_path}"
  cd "${func_path}" && \
    mv "${tmp_dir}/${func_name}" "${func_path}/${func_name}" && \
      chmod 600 "${func_path}/${func_name}"

  test -d "${minio_client_path}" || mkdir "${minio_client_path}"
  cd "${minio_client_path}" && \
    mv "${tmp_dir}/${minio_client_bin}" "${minio_client_path}/${minio_client_bin}" && \
      chmod 755 "${minio_client_path}/${minio_client_bin}"
}

test "$(id -u)" != "0" && { show_error "Sorry, but you must run this script as root."; exit 1; }

tmp_dir="${tmpdir}/lxd-functions.$$"
mkdir -p "${tmp_dir}" && \
  fetch_files "${tmp_dir}" && \
    install_files "${tmp_dir}" && \
        {
          test -x "${yq_bin_path}/${yq_bin}" && curl -s https://oss.help/scripts/tools/yq/update.sh | bash
          test -x "${yq_bin_path}/${yq_bin}" || curl -s https://oss.help/scripts/tools/yq/install.sh | bash
          show_notice "Minio-client was installed to ${minio_client_path}."
          show_notice "Library lxd-functions.sh (v${lxdf_ver}) was updated."
          show_notice "Script lxhelper (v${lxdh_ver}) was updated."
        }
test -d "${tmp_dir}" && rm -rf "${tmp_dir}"
test "${err}" -eq 1 && { show_error "Installation failed."; }
exit "${err}"
